consumers/campaigner/main.go

```diff
@@ -67,127 +91,121 @@ func (h *MessageHandler) HandleMessage(m *nsq.Message) error {
-
-				if s.Name != "" {
-					m["name"] = s.Name
-				}
-
-				url, err := s.GetUnsubscribeURL(msg.UserUUID)
-				if err != nil {
-					logrus.WithError(err).
-						WithField("subscriber", s).
-						Error("Unable to get unsubscribe url.")
-				} else {
-					m["unsubscribe_url"] = url
-				}
-
-				jsonMeta, err := json.Marshal(m)
-				if err != nil {
-					logrus.WithError(err).
-						WithField("subscriber", s).
-						Error("Unable to marshal metadata to json.")
-
-					continue
-				}
-				d := &ses.BulkEmailDestination{
-					Destination: &ses.Destination{
-						ToAddresses: []*string{aws.String(s.Email)},
-					},
-					ReplacementTemplateData: aws.String(string(jsonMeta)),
-				}
-
-				dest = append(dest, d)
-			}
+		template, err := h.svc.GetTemplate(context.Background(), msg.TemplateID, msg.UserID)
+		if err != nil {
+			logrus.WithError(err).
+				WithField("template_id", msg.TemplateID).
+				Error("unable to get template")
+			return nil
+		}
+		html, err := mustache.ParseString(template.HTMLPart)
+		if err != nil {
+			logrus.WithError(err).
+				WithField("html_part", template.HTMLPart).
+				Error("unable to parse html_part")
+			return nil
+		}
+		txt, err := mustache.ParseString(template.SubjectPart)
+		if err != nil {
+			logrus.WithError(err).
+				WithField("subject_part", template.SubjectPart).
+				Error("unable to parse subject_part")
+			return nil
+		}
+		sub, err := mustache.ParseString(template.TextPart)
+		if err != nil {
+			logrus.WithError(err).
+				WithField("text_part", template.TextPart).
+				Error("unable to parse text_part")
+			return nil
+		}
 
-			defaultData, err := json.Marshal(msg.TemplateData)
+		for _, s := range subs {
+			m, err := s.GetMetadata()
 			if err != nil {
-				logrus.WithError(err).Error("Unable to marshal template data as JSON.")
+				logrus.WithError(err).
+					WithField("subscriber", s).
+					Error("Unable to get subscriber metadata.")
 				continue
 			}
+			//  todo merge def template data with sub metadata
 
-			// prepare message for publishing to the queue
-			input := &ses.SendBulkTemplatedEmailInput{
-				Source:              aws.String(msg.Source),
-				Template:            aws.String(msg.Campaign.Template.Name),
-				Destinations:        dest,
-				DefaultTemplateData: aws.String(string(defaultData)),
-				DefaultTags: []*ses.MessageTag{
-					&ses.MessageTag{
-						Name:  aws.String("campaign_id"),
-						Value: aws.String(strconv.FormatInt(msg.Campaign.ID, 10)),
-					},
-					&ses.MessageTag{
-						Name:  aws.String("user_id"),
-						Value: aws.String(msg.UserUUID),
-					},
-				},
-			}
+			// fill template buffer with rendered template.
+			var (
+				htmlBuf bytes.Buffer
+				subBuf  bytes.Buffer
+				textBuf bytes.Buffer
+			)
 
-			if msg.ConfigurationSetExists {
-				input.ConfigurationSetName = aws.String(emails.ConfigurationSetName)
+			err = html.FRender(&htmlBuf, m)
+			if err != nil {
+				logrus.WithError(err).Error("unable to render html_part")
```

Comment:
Add fields here as well. Also not sure how we'll handle this one, it's a bit tricky. We shouldn't send an email to a subscriber if we don't render the subject and content properly. We need to discuss this further.